### ISSUE TITLE
Bandpass not read correctly with some miriad data sets

### ISSUE THREAD

**Issue by epaell** [HERA-Team/pyuvdata#123]
Hi All,

With updates to the ATNF flavour of miriad many years back, datasets can have time-based bandpass solutions and the bandpass file has an additional double float representing the solution time as a Julian date at the end of the file. When these datasets are read with aipy.miriad the bandpass that is read has an extra entry that causes problems when attempting to reshape it for the appropriate dimensions associated with the data set.

This can be demonstrated with the following script:

```
#!/usr/bin/env python2.7
import aipy
import numpy as np
import struct
import sys
fin = "uv.miriad"

# Use aipy to read in the basic data set parameters
uvdata = aipy.miriad.UV(fin)
nchan = uvdata['nchan']
nants = uvdata['nants']
nfeeds = uvdata['nfeeds']
nsols = uvdata['nsols']

# Print them out
print("nsols=%d; nchan=%d; nfeeds=%d; nants=%d" %(nsols, nchan, nfeeds, nants))
bp = uvdata['bandpass']
print("bandpass size=%d" %(bp.shape))
print("nchan*nfeeds*nants=%d" %(nchan*nants*nfeeds))
# The last value in the bandpass is actually the solution time as a Julian date (it is stored if nsols>1)
x=bp[nchan*nants*nfeeds]
ba = bytearray(struct.pack(">f", x.real)) + bytearray(struct.pack(">f", x.imag))
tcal = struct.unpack(">d", ba)
print("Julian date = %f" %(tcal))
```

When this is run with a data set calibrated with an ATNF miriad version of mfcal the following results:

**$ ./readbp.py** 
nsols=1; nchan=2049; nfeeds=2; nants=6
bandpass size=24589
nchan*nfeeds*nants=24588
Julian date = 2457080.662557

The bandpass itself has a total of 24588 entries but the aipy bandpass has 24589 entries because it inadvertently read the last double float as an additional entry. In fact, the last double float is a Julian date (which I recover in latter part of the Python script with some dodgy manipulation). The true bandpass should only be nchan*nfeeds*nants in length - the last double float should be ignored if present.

Cheers,

Emil.

**Comment by bhazelton** [comment]
Emil, can you point us at a file that we can use to test our development on this? A small test file that exhibited this behavior would be even better because then we could fold it into our unit tests.

**Comment by epaell** [comment]
Bryna, would the attached be suitable? It contains a simple ATCA data set (~1.2 MB) that has been calibrated and a simple script to read the bandpass via aipy (sorry, I just realised that there was extra code in the uploaded script that probably doesn't need to be there).

BTW, I contacted the developer with regards to the format of the bandpass file and apparently my interpretation of the file was correct, the last double float in the file is a time stamp.

[bptest.zip](https://github.com/HERA-Team/pyuvdata/files/791401/bptest.zip)

**Comment by bhazelton** [comment]
Thanks, Emil, that's very helpful. I'm trying to figure out how to integrate this with pyuvdata and I'm not a real Miriad user, so I have some questions:

How are the number of feeds and number of polarizations related? pyuvdata currently doesn't reference the number of feeds at all but it does use the number of polarizations, which your code doesn't, so I'm trying to figure out how to integrate this. The documentation we're working from lists both nfeeds and npol but doesn't discuss the relationship.

Do you have some documentation for this format? The document we've been working from lists nsol as a header item but it gives very little explanation, saying only that it is an integer representing "number of records in ‘gains’", which sounds a little different than what you described.

I'm aware that there are multiple miriad styles and we're happy to support as many as are needed by our users, but we have struggled to find good documentation of the various styles.

**Comment by epaell** [comment]
Hi Bryna,

The number of feeds relates to the number of receiving elements. So for example, with the MWA it would be 2 i.e. X and Y. The polarisations include the cross terms and so you have XX, YY, XY, and YX - 4 terms. My guess would be that if you only had a single dipole then nfeed would be the same as npol.

Sadly, the documentation appears to be non-existent - I had to contact one of the developers and he forwarded an earlier email he received from a previous developer ;-) The definition you mentioned from nsol seems consistent with my earlier understanding ie. if there is more than one solution in time then nsol will be greater than 1. In general there is only one solution (I'm not sure that I've ever encountered one with more than one but presumably it is possible given that they have provided a means to support it in the data format). However, I just read the developers email again and realised that he did state:

> The 8 bytes at the end are indeed a time field, only useful when the
> bandpass solution is time dependent (nbsols >1).

So now I'm not sure if nbsols is different to nsol or whether it was a typo in the email :-(

**Comment by bhazelton** [comment]
Hi Emil,

Thanks, I think I'm starting to get the idea. I'm a little confused, though, because talking about a single pol usually indicates that we're talking about a calibration solution rather than a visibility. Is that the case here?

If so, you should know that we are currently developing the object to store calibration solutions, but it's not yet part of the master branch (to see details, feel free to check out the cal branch). Once the calibration solution object exists, there's a plan to make it be able to read from and write to miriad cal structures, but no one has been identified to do that yet, so it may take some time.

**Comment by epaell** [comment]
Hi Bryna,

Yes, we are talking about a calibration solution - that is what is kept inside the bandpass file (the visibility data is kept in a separate file). Being able to read and write to the miriad cal structures would be absolutely fantastic (it was actually the reason why I was looking in to the details of the bandpass file in the first place, something Randall was also interested in).

**Comment by bhazelton** [comment]
Hi Emil,

Thanks for the clarification. I'm going to label this issue as relating to calibration and we will incorporate it in our miriad calibration code when we implement it. Are you interested in either helping to implement it or helping to test it once the general calibration structure is in place?

Thanks!
Bryna

**Comment by bhazelton** [comment]
This is addressed by the feature request in issue #107
